encoding/ccf/decode.go

```diff
@@ -0,0 +1,1887 @@
+	// Decode array length.
+	n, err := d.dec.DecodeArrayHead()
+	if err != nil {
+		return nil, err
+	}
+
+	if hasKnownSize && uint64(knownSize) != n {
+		return nil, fmt.Errorf(
+			"encoded array-value has %d elements (expected %d elements)",
+			n,
+			knownSize,
+		)
+	}
+
+	elementType := typ.Element()
+
+	values := make([]cadence.Value, n)
+	for i := 0; i < int(n); i++ {
+		// Decode value.
+		element, err := d.decodeValue(elementType, types)
+		if err != nil {
+			return nil, err
+		}
+		values[i] = element
+	}
+
+	v, err := cadence.NewMeteredArray(
+		d.gauge,
+		len(values),
+		func() ([]cadence.Value, error) {
+			return values, nil
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return v.WithType(typ), nil
+}
+
+// decodeDictionary decodes encoded dict-value as
+// language=CDDL
+// dict-value = [* (key: value, value: value)]
+func (d *Decoder) decodeDictionary(typ *cadence.DictionaryType, types cadenceTypeByCCFTypeID) (cadence.Value, error) {
+	// Decode array length.
+	n, err := d.dec.DecodeArrayHead()
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if number of elements is even.
+	if n%2 != 0 {
+		return nil, fmt.Errorf(
+			"encoded dict-value has %d elements (expected even number of elements)",
+			n,
+		)
+	}
+
+	// previousKeyRawBytes is used to determine if dictionary keys are sorted
+	var previousKeyRawBytes []byte
+
+	pairCount := n / 2
+	pairs := make([]cadence.KeyValuePair, pairCount)
+	for i := 0; i < int(pairCount); i++ {
+		// element i: key
+
+		// Decode key as raw bytes to check that key pairs are sorted by key.
+		keyRawBytes, err := d.dec.DecodeRawBytes()
+		if err != nil {
+			return nil, err
+		}
+
+		// "Deterministic CCF Encoding Requirements" in CCF specs:
+		//
+		//   "dict-value key-value pairs MUST be sorted by key."
+		if !bytesAreSortedBytewise(previousKeyRawBytes, keyRawBytes) {
+			return nil, fmt.Errorf("encoded dict-value keys are not sorted")
+		}
+
+		previousKeyRawBytes = keyRawBytes
+
+		// decode key from raw bytes
+		keyDecoder := NewDecoder(d.gauge, keyRawBytes)
+		key, err := keyDecoder.decodeValue(typ.KeyType, types)
+		if err != nil {
+			return nil, err
+		}
+
+		// element i+1: value
+		element, err := d.decodeValue(typ.ElementType, types)
+		if err != nil {
+			return nil, err
+		}
+
+		pairs[i] = cadence.NewMeteredKeyValuePair(d.gauge, key, element)
+	}
+
+	value, err := cadence.NewMeteredDictionary(
+		d.gauge,
+		len(pairs),
+		func() ([]cadence.KeyValuePair, error) {
+			// "Valid CCF Encoding Requirements" in CCF specs:
+			//
+			//   "Keys MUST be unique in dict-value. Decoders are not always required to check
+			//   for duplicate dictionary keys. In some cases, checking for duplicate dictionary
+			//   key is not necessary or it may be delegated to the application."
+			//
+			// Here, decoder doesn't check uniqueness of dictionary keys
+			// because checking is delegated (entrusted) to Cadence runtime.
+			return pairs, nil
```

Comment:
The allocation of `[]cadence.KeyValuePair` should occur in this closure, so that memory is metered before the memory is actually allocated

Reply:
Good catch!  Done in commit 2e6e4a4cd53ad48abd3ae5f12080656cceca4c9b.